pkg/reconciler/openshift/tektonpipeline/installerset.go

```diff
@@ -35,58 +36,72 @@ const createdByValue = "TektonPipeline"
 // and if installer set which already exist is of older version/ or if target namespace is different then it
 // deletes and return false to create a new installer set
 func checkIfInstallerSetExist(ctx context.Context, oc clientset.Interface, relVersion string,
-	tp *v1alpha1.TektonPipeline, component string) (bool, error) {
+	tp *v1alpha1.TektonPipeline, ls string) (bool, error) {
 
 	// Check if installer set is already created
-	compInstallerSet, ok := tp.Status.ExtentionInstallerSets[component]
-	if !ok {
+	installerSets, err := oc.OperatorV1alpha1().TektonInstallerSets().
+		List(ctx, metav1.ListOptions{
+			LabelSelector: ls,
+		})
+	if err != nil {
+		return false, err
+	}
+	if len(installerSets.Items) == 0 {
+		// only scenario where this function returns
+		// false, nil
 		return false, nil
 	}
 
-	if compInstallerSet != "" {
-		// if already created then check which version it is
-		ctIs, err := oc.OperatorV1alpha1().TektonInstallerSets().
-			Get(ctx, compInstallerSet, metav1.GetOptions{})
+	// if the List query returns more than 1 InstallerSets,
+	// then delete all of them as we are expecting only one
+	// If there are more than one then it means the state is
+	// corrupted/unexpected. So to get back to desired state
+	// delete all installerSets that match the labelSelector
+	if len(installerSets.Items) > 1 {
+		err = oc.OperatorV1alpha1().TektonInstallerSets().
+			DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{
+				LabelSelector: ls,
+			})
+
 		if err != nil {
-			if errors.IsNotFound(err) {
-				return false, nil
-			}
 			return false, err
 		}
+		return false, v1alpha1.RECONCILE_AGAIN_ERR
+	}
 
-		// Check release version and target namespace
-		// If anyone of this is not as expected then delete existing
-		// installer set and create a new one
+	// if the InstallerSet already exists then check then validate it:
+	// Check release version and target namespace
+	// If anyone of this is not as expected then delete existing
+	// InstallerSet and return false
```

Comment:
@nikhil-thomas 
can we do 
if len(installerSets.Items) == 1 {
// check version n namespace match
// If matched return true
}

// Delete installer sets like below
 oc.OperatorV1alpha1().TektonInstallerSets().
		DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{
			LabelSelector: labelSelector,
		})
// which covers when len is 0, >1 and version/ns mismatch scenario

WDYT?

Reply:
We could do that if the response of `>1` and `==0`were same.
but, here are double checking that the installer set is completely deleted even after no error from DeleteCollection API call.

if we don't do this then deleteCollection call will add DeleteTimeStamps and infact resources could linger around when the next new InstallerSet of same type comes up.

As mentioned in the `// only scenario where this function returns // false, nil` comment. 
this function returns `false, nil` (so that the caller could continue) only if there are 0 installerSets on cluster.

In `>1` scenario,  DeleteCollection() with no error, the function returns `Reconcile Again error` so that the the function can ensure in the next loop that the state is indeed no: of installerSets == 0.